libkbfs/folder_branch_ops.go

```diff
@@ -4593,6 +4605,66 @@ func (fbo *folderBranchOps) unstageAfterFailedResolution(ctx context.Context,
 	return fbo.unstageLocked(ctx, lState)
 }
 
+func (fbo *folderBranchOps) onTLFBranchChange(newBID BranchID) {
+	ctx, cancelFunc := fbo.newCtxWithFBOID()
+	defer cancelFunc()
+
+	// This only happens on a `PruneBranch` call, in which case we
+	// would have already updated fbo's local view of the branch/head.
+	if newBID == NullBranchID {
+		fbo.log.CDebugf(ctx, "Ignoring branch change back to master")
+		return
+	}
+
+	lState := makeFBOLockState()
+	fbo.mdWriterLock.Lock(lState)
+	defer fbo.mdWriterLock.Unlock(lState)
+
+	fbo.log.CDebugf(ctx, "Journal branch change: %s", newBID)
+
+	if !fbo.isMasterBranchLocked(lState) {
+		if fbo.bid == newBID {
+			fbo.log.CDebugf(ctx, "Already on branch %s", newBID)
+			return
+		}
+		panic(fmt.Sprintf("Cannot switch to branch %s while on branch %s",
+			newBID, fbo.bid))
+	}
+
+	md, err := fbo.config.MDOps().GetUnmergedForTLF(ctx, fbo.id(), newBID)
+	if err != nil {
+		fbo.log.CWarningf(ctx,
+			"No unmerged head on journal branch change (bid=%s)", newBID)
+		return
+	}
+
+	if md == (ImmutableRootMetadata{}) || md.MergedStatus() != Unmerged ||
```

Comment:
This can happen in practice, right? If we convert to a branch and then quickly flush everything, and then CR happens. May be worth a comment.